src/drivers/imx/esai.c

```diff
@@ -4,37 +4,456 @@
+
+	/* Set networked mode; we only support 2 channels now, not 1 */
+	xcr |= ESAI_xCR_xMOD_NETWORK;
+	xccr |= ESAI_xCCR_xDC(2);
+
+	/* Word size -- TODO compute correct one */
+	xcr |= ESAI_xCR_xSWS(32, 24) | ESAI_xCR_PADC;
+
+	/* Remove "RESET" flag */
+	dai_update_bits(dai, REG_ESAI_ECR, ESAI_ECR_ERST, 0);
+	
+	dai_update_bits(dai, REG_ESAI_ECR, ESAI_ECR_ETI, ESAI_ECR_ETI);
+
+	mask = ESAI_xCCR_xCKP | ESAI_xCCR_xHCKP | ESAI_xCCR_xFSP |
+		ESAI_xCCR_xFSD | ESAI_xCCR_xCKD | ESAI_xCCR_xHCKD |
+		ESAI_xCCR_xDC_MASK;
+
+	xccr |= ESAI_xCCR_xHCKD; /* Set the HCKT pin as an output */
+
+	if (~mask & xccr) {
+		trace_esai_error("XCCR bits not caught by mask: 0x%08x",
+				 (~mask & xccr));
+		trace_esai_error("MASK 0x%08x XCCR 0x%08x", mask, xccr);
+	}
+
+	dai_update_bits(dai, REG_ESAI_TCCR, mask, xccr); /* rx */
+	dai_update_bits(dai, REG_ESAI_RCCR, mask, xccr); /* tx */
+
+	mask = ESAI_xCR_xFSL | ESAI_xCR_xFSR | ESAI_xCR_xWA |
+		ESAI_xCR_xMOD_MASK | ESAI_xCR_xSWS_MASK | ESAI_xCR_PADC |
+		ESAI_xCR_xPR;
+	xcr |= ESAI_xCR_xPR;
+
+	if (~mask & xcr) {
+		trace_esai_error("XCR bits not caught by mask: 0x%08x",
+				 (~mask & xcr));
+		trace_esai_error("MASK 0x%08x XCR 0x%08x", mask, xcr);
+	}
+
+	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr); /* rx */
+	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr); /* tx */
+
+	dai_write(dai, REG_ESAI_TSMA, 0);
+	dai_write(dai, REG_ESAI_TSMB, 0);
+	dai_write(dai, REG_ESAI_RSMA, 0);
+	dai_write(dai, REG_ESAI_RSMB, 0);
+
+	/* Program FIFOs */
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, 0);
+	
+	/* TODO use TFCR and RFCR */
+	dai_update_bits(dai, REG_ESAI_TFCR,
+				ESAI_xFCR_xFR_MASK,
+				ESAI_xFCR_xFR);
+
+	dai_update_bits(dai, REG_ESAI_TFCR,
+				ESAI_xFCR_xFR_MASK | ESAI_xFCR_xWA_MASK | ESAI_xFCR_xFWM_MASK | ESAI_xFCR_TE_MASK | ESAI_xFCR_TIEN,
+				ESAI_xFCR_xWA(24) | ESAI_xFCR_xFWM(96) | ESAI_xFCR_TE(1) | ESAI_xFCR_TIEN);
+	/* TODO */
+	/* ESAI_xFCR_xWA(bits) instead of 16 ^^^^ */
+	/* ESAI_xFCR_xFWM(slots) -- compute slots instead of hardcoding
+	 * 32!
+	 */
+
+	/* Set the clock divider to divide EXTAL by 16 (DIV8 from PSR,
+	 * plus a divide by 2 which is mandatory overall)
+	 */
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xDC_MASK, ESAI_xCCR_xDC(2));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xFP_MASK, ESAI_xCCR_xFP(1));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xPSR_MASK, ESAI_xCCR_xPSR_DIV8);
+
+
+	/* Remove ESAI personal reset */
+	dai_update_bits(dai, REG_ESAI_xCR(0), ESAI_xCR_xPR, 0);
+	dai_update_bits(dai, REG_ESAI_xCR(1), ESAI_xCR_xPR, 0);
+
+	dai_update_bits(dai, REG_ESAI_PRRC, ESAI_PRRC_PDC_MASK,
+			ESAI_PRRC_PDC(ESAI_GPIO));
+	dai_update_bits(dai, REG_ESAI_PCRC, ESAI_PCRC_PC_MASK,
+			ESAI_PCRC_PC(ESAI_GPIO));
+
+	tracev_esai("ESAI_REGS_DUMP in esai_set_config");
+	esai_regs_dump(dai);
 	return 0;
 }
 
+static void esai_start(struct dai *dai, int direction)
+{
+	direction = direction == SOF_IPC_STREAM_PLAYBACK;
```

Comment:
I think we've agreed in different PR about direct using of direction parameter.

Reply:
I think I fixed this.